### PR TITLE
Adr 764: Product List screen

### PR DESCRIPTION
**Pull request by msatta**

### CODE REVIEW

File: app/controllers/productEntry/CheckYourAnswersController.scala [hunk 1]
```diff
@@ -43,30 +46,57 @@ class CheckYourAnswersController @Inject() (
...
-
     }
+  }
 
+  private def getProductEntry(answers: UserAnswers, maybeInt: Option[Int]): Option[ProductEntry] = {
+    val pe = maybeInt match {
+      case Some(i) => answers.getByIndex(ProductEntryListPage, i)
+      case None    => answers.get(CurrentProductEntryPage)
+    }
+    pe.map(pe => if (pe.index.isDefined) pe else pe.copy(index = maybeInt))
```
Comment (yasaracar) [inline app/controllers/productEntry/CheckYourAnswersController.scala:82]: Why do we do this? Isn't maybeInt None anyways if it is a CurrentProductEntry?

File: app/controllers/productEntry/CheckYourAnswersController.scala [hunk 2]
```diff
@@ -43,30 +46,57 @@ class CheckYourAnswersController @Inject() (
...
   }
 
+  private def saveProductEntry(
+    answers: UserAnswers,
+    productEntry: ProductEntry
+  ): Future[UserAnswers] =
+    productEntry.index match {
+      case Some(i) => Future.fromTry(answers.setByIndex(ProductEntryListPage, i, productEntry.copy(index = None)))
```
Comment (yasaracar) [inline app/controllers/productEntry/CheckYourAnswersController.scala:90]: Why do we set the index to None?

File: app/models/UserAnswers.scala [hunk 3]
```diff
@@ -94,6 +94,16 @@ final case class UserAnswers(
       case JsError(errors)       =>
         Failure(JsResultException(errors))
     }
+
+  def setByIndex[A](page: Settable[Seq[A]], i: Int, value: A)(implicit writes: Writes[A]): Try[UserAnswers] = {
```
Comment (yasaracar) [inline app/models/UserAnswers.scala:98]: rename i to index please

File: app/navigation/ProductEntryNavigator.scala [hunk 4]
```diff
@@ -69,4 +70,11 @@ class ProductEntryNavigator @Inject() () extends BaseNavigator {
       case Some(false) => routes.IndexController.onPageLoad
       case _           => routes.JourneyRecoveryController.onPageLoad()
     }
+
+  private def productListPageRoute(answers: UserAnswers): Call =
```
Comment (yasaracar) [inline app/navigation/ProductEntryNavigator.scala:74]: Test these please

**Review by yasaracar — CHANGES_REQUESTED** [review]

File: app/controllers/productEntry/CheckYourAnswersController.scala (hunk 1 above)
Comment (sini-george) [reply app/controllers/productEntry/CheckYourAnswersController.scala:82]: maybeInt can have an index when user comes to the CYA page from clicking on 'change' on the Product List page. This index will be stored to the CurrentProductEntry. It helps when we need to decide whether we need to change a product entry in the list or the add a new one.

File: app/controllers/productEntry/CheckYourAnswersController.scala (hunk 2 above)
Comment (sini-george) [reply app/controllers/productEntry/CheckYourAnswersController.scala:90]: This is to clear the index we stored in the CurrentProductEntry when it comes from the 'Change' case described above

**Review by yasaracar — DISMISSED** [review]

**Review by yasaracar — APPROVED** [review]

**Pull request merged**